File: blog-backend/src/main/java/com/blogproject/blogproject/service/SocialMediaService.java
package com.blogproject.blogproject.service;

import com.blogproject.blogproject.dtos.SocialMediaCreation;
import com.blogproject.blogproject.dtos.SocialMediaDTO;
import com.blogproject.blogproject.entities.SocialMedia;
import com.blogproject.blogproject.entities.User;
import com.blogproject.blogproject.enums.SocialMediaType;
import com.blogproject.blogproject.repository.SocialMediaRepository;
import com.blogproject.blogproject.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Slf4j
@Transactional
public class SocialMediaService {

    private final SocialMediaRepository socialMediaRepository;
    private final UserRepository userRepository;


    // =========================================================
    // MAPPING
    // =========================================================

    private SocialMediaDTO mapToDTO(
            SocialMedia socialMedia,
            User user
    ) {

        SocialMediaDTO dto = new SocialMediaDTO();

        dto.setId(socialMedia.getId());
        dto.setName(socialMedia.getName());
        dto.setLink(socialMedia.getLink());
        dto.setDescription(socialMedia.getDescription());
        dto.setSocialMediaType(socialMedia.getType());

        dto.setUserName(user.getName());
        dto.setUserEmail(user.getEmail());

        return dto;
    }


    // =========================================================
    // USER
    // =========================================================

    private User getUserByEmail(String email) {

        return userRepository.findByEmail(email)
                .orElseThrow(() ->
                        new RuntimeException("User not found"));
    }


    // =========================================================
    // CREATE
    // =========================================================

    public SocialMediaDTO saveSocialMedia(
            SocialMediaCreation creation,
            String email
    ) {

        User user = getUserByEmail(email);

        if (socialMediaRepository.existsByLink(creation.getLink())) {
            throw new RuntimeException(
                    "This social media link is already used"
            );
        }

        SocialMedia socialMedia =
                SocialMedia.builder()
                        .name(creation.getName())
                        .link(creation.getLink())
                        .description(creation.getDescription())
                        .type(creation.getSocialMediaType())
                        .userId(user.getId())
                        .build();

        SocialMedia saved =
                socialMediaRepository.save(socialMedia);

        return mapToDTO(saved, user);
    }


    // =========================================================
    // GET BY ID
    // =========================================================

    @Transactional(readOnly = true)
    public SocialMediaDTO getSocialMediaById(
            String id
    ) {

        SocialMedia socialMedia =
                socialMediaRepository.findById(id)
                        .orElseThrow(() ->
                                new RuntimeException(
                                        "Social media not found"
                                ));

        User user =
                userRepository.findById(socialMedia.getUserId())
                        .orElseThrow(() ->
                                new RuntimeException(
                                        "User not found"
                                ));

        return mapToDTO(socialMedia, user);
    }


    // =========================================================
    // GET BY PAGE
    // =========================================================

    @Transactional(readOnly = true)
    public Page<SocialMediaDTO> getSocialMediaByPage(
            int page,
            int size,
            String email
    ) {

        User user = getUserByEmail(email);

        Pageable pageable =
                PageRequest.of(page, size);

        return socialMediaRepository
                .findByUserId(
                        user.getId(),
                        pageable
                )
                .map(socialMedia ->
                        mapToDTO(socialMedia, user)
                );
    }


    // =========================================================
    // GET BY TYPE
    // =========================================================

    @Transactional(readOnly = true)
    public Page<SocialMediaDTO> getSocialMediaByType(
            int page,
            int size,
            SocialMediaType type,
            String email
    ) {

        User user = getUserByEmail(email);

        Pageable pageable =
                PageRequest.of(page, size);

        return socialMediaRepository
                .findByUserIdAndType(
                        user.getId(),
                        type,
                        pageable
                )
                .map(socialMedia ->
                        mapToDTO(socialMedia, user)
                );
    }


    // =========================================================
    // UPDATE
    // =========================================================

    public SocialMediaDTO updateSocialMediaById(
            String id,
            SocialMediaDTO dto,
            String email
    ) {

        User user = getUserByEmail(email);

        SocialMedia socialMedia =
                socialMediaRepository.findById(id)
                        .orElseThrow(() ->
                                new RuntimeException(
                                        "Social media not found"
                                ));

        /*
         * Make sure the user can only modify
         * his own social media.
         */
        if (!socialMedia.getUserId().equals(user.getId())) {
            throw new RuntimeException(
                    "Unauthorized access"
            );
        }

        /*
         * If the link changes, make sure the new
         * link isn't already used.
         */
        if (!socialMedia.getLink().equals(dto.getLink())
                && socialMediaRepository.existsByLink(dto.getLink())) {

            throw new RuntimeException(
                    "This social media link is already used"
            );
        }

        socialMedia.setName(dto.getName());
        socialMedia.setLink(dto.getLink());
        socialMedia.setDescription(dto.getDescription());
        socialMedia.setType(dto.getSocialMediaType());

        SocialMedia updated =
                socialMediaRepository.save(socialMedia);

        return mapToDTO(updated, user);
    }


    // =========================================================
    // DELETE
    // =========================================================

    public void deleteSocialMediaById(
            String id,
            String email
    ) {

        User user = getUserByEmail(email);

        SocialMedia socialMedia =
                socialMediaRepository.findById(id)
                        .orElseThrow(() ->
                                new RuntimeException(
                                        "Social media not found"
                                ));

        if (!socialMedia.getUserId().equals(user.getId())) {
            throw new RuntimeException(
                    "Unauthorized access"
            );
        }

        socialMediaRepository.delete(socialMedia);
    }


    // =========================================================
    // CHECK LINK
    // =========================================================

    @Transactional(readOnly = true)
    public boolean isLinkUsed(
            String link
    ) {

        return socialMediaRepository.existsByLink(link);
    }
}

File: blog-frontend/src/app/user-profile/service/social-media.service.ts
import { Injectable } from '@angular/core';
import {HttpClient, HttpParams} from "@angular/common/http";
import {SocialMedia, SocialMediaCreation} from "../../models/SocialMedia";
import {Observable} from "rxjs";

@Injectable({
  providedIn: 'root',
})
export class SocialMediaService {
  private readonly BASE_URL = 'http://localhost:8083/api/socials';

  constructor(private http: HttpClient) {}

  createSocialMedia(socialMediaCreation: SocialMediaCreation, email: string): Observable<SocialMedia> {
    const params = new HttpParams().set('email', email);
    return this.http.post<SocialMedia>(`${this.BASE_URL}`, socialMediaCreation, { params });
  }

  getSocialMediaById(id: string): Observable<SocialMedia> {
    return this.http.get<SocialMedia>(`${this.BASE_URL}/${id}`);
  }

  deleteSocialMediaById(id: any, email: string): Observable<void> {
    const params = new HttpParams().set('email', email);
    return this.http.delete<void>(`${this.BASE_URL}/${id}`, { params });
  }

  getAllSocialMedia(page: number, size: number, email: string): Observable<any> {
    const params = new HttpParams()
      .set('page', page)
      .set('size', size)
      .set('email', email);
    return this.http.get<any>(`${this.BASE_URL}`, { params });
  }

  updateSocialMediaById(id: any, socialMediaDTO: SocialMedia, email: string): Observable<void> {
    const params = new HttpParams().set('email', email);
    return this.http.put<void>(`${this.BASE_URL}/${id}`, socialMediaDTO, { params });
  }

  isLinkUsed(link: string): Observable<boolean> {
    const params = new HttpParams().set('link', link);
    return this.http.get<boolean>(`${this.BASE_URL}/link/used`, { params });
  }

  getSocialMediaByType(page: number, size: number, type: string, email: string): Observable<any> {
    const params = new HttpParams()
      .set('page', page)
      .set('size', size)
      .set('email', email);
    return this.http.get<any>(`${this.BASE_URL}/type/${type}`, { params });
  }
}
